### PR TITLE
feat: Web 160 configure longer redux multicall blocks per fetch for l2 chains

### PR DESCRIPTION
**Pull request by lynnshaoyu**
Add ability to configure default ListenerOptions as a parameter in each Updater. blocksPerFetch is a param in default listener options, and being able to set this at the Updater level (instead of only hooks call level as before) enables us to configure blocksPerFetch differently for different chains. Introducing this in response to request to lower frequency of multicall calls for L2 chains (that can create new blocks every second for example).

### CODE REVIEW

File: src/slice.ts [hunk 1]
```diff
@@ -97,6 +103,15 @@ export function createMulticallSlice(reducerPath: string) {
           }
         })
       },
+
+      updateDefaultListenerOptions: (state, action: PayloadAction<MulticallListenerOptionsPayload>) => {
+        const { chainId, listenerOptions } = action.payload
+        console.log(`updateDefaultListenerOptions with ${chainId} and ${listenerOptions}`)
```
Comment (JFrankfurt) [inline src/slice.ts:109]: can drop the log

File: package.json [hunk 2]
```diff
@@ -51,7 +61,7 @@
     "@ethersproject/bignumber": "5",
     "@ethersproject/contracts": "5",
     "@reduxjs/toolkit": "1",
-    "react": "17",
+    "react": "^16.14.0",
```
Comment (JFrankfurt) [inline package.json:64]: what makes these necessary?

Comment (lynnshaoyu) [reply package.json]: you're right its not, accidentally changed during my testing experimentations. reverted

File: package.json [hunk 3]
```diff
@@ -32,8 +32,9 @@
     "@reduxjs/toolkit": "^1.6.2",
     "@testing-library/jest-dom": "^5.15.1",
     "@testing-library/react": "^12.1.2",
+    "@testing-library/react-hooks": "^8.0.0",
     "@typechain/ethers-v5": "^7.2.0",
-    "@types/jest": "^24.0.25",
+    "@types/jest": "^25.2.1",
     "@types/react": "^17.0.28",
     "@types/react-dom": "^17.0.14",
     "@uniswap/v3-periphery": "^1.3.0",
```
Comment (zzmp) [inline package.json:40]: the react peerDep should be made >=17

File: src/types.ts [hunk 4]
```diff
@@ -47,6 +47,10 @@ export interface MulticallState {
       }
     }
   }
+
+  defaultListenerOptions?: {
```
Comment (zzmp) [inline src/types.ts:51]: strike `default` - it's a user-setting, not a default setting

File: src/slice.ts [hunk 5]
```diff
@@ -97,6 +103,14 @@ export function createMulticallSlice(reducerPath: string) {
           }
         })
       },
+
+      updateDefaultListenerOptions: (state, action: PayloadAction<MulticallListenerOptionsPayload>) => {
```
Comment (zzmp) [inline src/slice.ts:107]: strike `Default`

File: package.json [hunk 6]
```diff
@@ -32,8 +32,9 @@
     "@reduxjs/toolkit": "^1.6.2",
     "@testing-library/jest-dom": "^5.15.1",
     "@testing-library/react": "^12.1.2",
+    "@testing-library/react-hooks": "^8.0.0",
```
Comment (JFrankfurt) [inline package.json:35]: it looks like you don't end up using this (it's commented out in the test)
drop it?

File: integration-tests/MultichainApp.tsx [hunk 7]
```diff
@@ -15,7 +15,7 @@ export function MultichainApp() {
   return (
     <Provider store={store}>
       <Updater chainId={ChainId.MAINNET} blockNumber={blockNumberMainnet} />
-      <Updater chainId={ChainId.RINKEBY} blockNumber={blockNumberRinkeby} />
+      <Updater chainId={ChainId.RINKEBY} blockNumber={blockNumberRinkeby} blocksPerFetch={5} />
       <Home chainIds={chains} blockNumbers={blocks} />
```
Comment (JFrankfurt) [inline integration-tests/MultichainApp.tsx:19]: Why rinkeby, but not mainnet here?

File: src/hooks.test.tsx [hunk 8]
```diff
@@ -124,5 +134,121 @@ describe('multicall hooks', () => {
...
+            <Provider store={store}>
+              <Caller calls={[callA]} multicallContext={mockContext} />
+            </Provider>,
+            container
+          )
+        })
+
+        expect(mockContext.actions.addMulticallListeners).toHaveBeenCalledWith({
+          chainId: 1,
+          calls: [{ address: 'a', callData: '' }],
```
Comment (JFrankfurt) [inline src/hooks.test.tsx:175]: nit: re-use the initial argument variable above

File: src/slice.ts [hunk 9]
```diff
@@ -97,6 +103,14 @@ export function createMulticallSlice(reducerPath: string) {
...
+        const { chainId, listenerOptions } = action.payload
+        const defaultListenerOptions: MulticallState['defaultListenerOptions'] = state.defaultListenerOptions
```
Comment (zzmp) [inline src/slice.ts:109]: Why do you need to assign a variable? Also, split out the defaulting to make it more readable:

```
state.listenerOptions = state.listenerOptions ?? {}
state.listenerOptions[chainId] = listenerOptions
```

File: src/hooks.test.tsx (hunk 8 above)
Comment (JFrankfurt) [reply src/hooks.test.tsx:175]: (same below)

File: src/hooks.ts [hunk 10]
```diff
@@ -16,24 +16,27 @@ export function useCallsDataSubscription(
...
-
   const serializedCallKeys: string = useMemo(() => JSON.stringify(callsToCallKeys(calls)), [calls])
 
   // update listeners when there is an actual change that persists for at least 100ms
   useEffect(() => {
     const callKeys: string[] = JSON.parse(serializedCallKeys)
     const calls = callKeysToCalls(callKeys)
     if (!chainId || !calls) return
+    const blocksPerFetchForChain =
+      blocksPerFetch ?? (defaultListenerOptions ?? {})[chainId]?.blocksPerFetch ?? DEFAULT_BLOCKS_PER_FETCH
```
Comment (JFrankfurt) [inline src/hooks.ts:33]: can you unpack this? the logic is nested enough that is starting to get hard to parse

File: src/slice.ts [hunk 11]
```diff
@@ -97,6 +103,14 @@ export function createMulticallSlice(reducerPath: string) {
...
+          ? state.defaultListenerOptions
+          : (state.defaultListenerOptions = {})
```
Comment (JFrankfurt) [inline src/slice.ts:111]: is this an assignment happening in the middle of the ternary? I think it should probably be unpacked

File: src/updater.tsx [hunk 12]
```diff
@@ -99,7 +100,7 @@ export function activeListeningKeys(
 
     memo[callKey] = Object.keys(keyListeners)
       .filter((key) => {
-        const blocksPerFetch = parseInt(key)
+        const blocksPerFetch = !!listenerOptions?.blocksPerFetch ? listenerOptions.blocksPerFetch : parseInt(key)
```
Comment (zzmp) [inline src/updater.tsx:103]: I think at this point, the blocksPerFetch should be encoded into the state, so that you do not need to pass `listenerOptions` into this function.

In other words, I think you can revert L91-92, L103, L248-249.

File: src/updater.tsx [hunk 13]
```diff
@@ -121,15 +122,16 @@ export function outdatedListeningKeys(
...
+  listenerOptions?: ListenerOptions
 ): string[] {
   if (!chainId || !latestBlockNumber) return []
   const results = callResults[chainId]
   // no results at all, load everything
   if (!results) return Object.keys(listeningKeys)
 
   return Object.keys(listeningKeys).filter((callKey) => {
-    const blocksPerFetch = listeningKeys[callKey]
+    const blocksPerFetch = !!listenerOptions?.blocksPerFetch ? listenerOptions.blocksPerFetch : listeningKeys[callKey]
```
Comment (zzmp) [inline src/updater.tsx:134]: I think at this point, the blocksPerFetch should be encoded into the state, so that you do not need to pass listenerOptions into this function.

In other words, I think you can revert L125-126, L134, L252-260, and src/updater.test.ts.

File: src/hooks.ts [hunk 14]
```diff
@@ -77,10 +80,11 @@ function areCallResultsEqual(a: CallResult[], b: CallResult[]) {
 function useMultichainCallsDataSubscription(
   context: MulticallContext,
   chainToCalls: Record<number, Array<Call | undefined>>,
-  blocksPerFetch = 1
+  blocksPerFetch?: number
```
Comment (zzmp) [inline src/hooks.ts:83]: Why not change this to `ListenerOptions`?

File: src/hooks.ts [hunk 15]
```diff
@@ -16,24 +16,27 @@ export function useCallsDataSubscription(
   context: MulticallContext,
   chainId: number | undefined,
   calls: Array<Call | undefined>,
-  blocksPerFetch = 1
+  blocksPerFetch?: number
```
Comment (zzmp) [inline src/hooks.ts:19]: Why not change this to `ListenerOptions`? Then you can pass that directly to the dispatcher, and deal with the defaults in the dispatch actions (in slice.ts).

**Review by zzmp — DISMISSED** [review]

File: src/types.ts (hunk 4 above)
Comment (lynnshaoyu) [reply src/types.ts:51]: done

File: src/slice.ts (hunk 5 above)
Comment (lynnshaoyu) [reply src/slice.ts:107]: done

File: src/slice.ts (hunk 11 above)
Comment (lynnshaoyu) [reply src/slice.ts:111]: Sounds good, done.

File: src/hooks.test.tsx (hunk 8 above)
Comment (lynnshaoyu) [reply src/hooks.test.tsx:175]: Sounds good done

File: integration-tests/Updater.tsx [hunk 16]
```diff
@@ -6,9 +6,20 @@ import { multicall } from './multicall'
...
   blockNumber: number | undefined
+  blocksPerFetch?: number
 }
 
-export function Updater({ chainId, blockNumber }: Props) {
+export function Updater({ chainId, blockNumber, blocksPerFetch }: Props) {
   const contract = useContract(chainId)
-  return <multicall.Updater chainId={chainId} latestBlockNumber={blockNumber} contract={contract} />
+  const listenerOptions = blocksPerFetch ? { blocksPerFetch } : undefined
+  return listenerOptions ? (
```
[the remark this answers is not in the document]
Comment (lynnshaoyu) [reply integration-tests/Updater.tsx:15]: Done

File: package.json (hunk 6 above)
Comment (lynnshaoyu) [reply package.json:35]: yep you're right, thanks for the catch. removed.

File: src/hooks.ts (hunk 10 above)
Comment (lynnshaoyu) [reply src/hooks.ts:33]: good call, done

File: src/hooks.ts (hunk 15 above)
Comment (lynnshaoyu) [reply src/hooks.ts:19]: Sgtm done

File: src/hooks.ts (hunk 14 above)
Comment (lynnshaoyu) [reply src/hooks.ts:83]: Sgtm done

File: integration-tests/MultichainApp.tsx (hunk 7 above)
Comment (lynnshaoyu) [reply integration-tests/MultichainApp.tsx:19]: Just testing that we can set blocksPerFetch to different values for different chains. I presume ETH will usually default to blocksPerFetch = 1 and other chains will be configured differently so this just simulates that.

File: package.json (hunk 3 above)
Comment (lynnshaoyu) [reply package.json:40]: They are already >=17 I believe?

File: src/updater.tsx (hunk 12 above)
Comment (lynnshaoyu) [reply src/updater.tsx:103]: You're right this was an important call out actually because I _should_ be getting blocksPerFetch directly from callListeners. What is happening here is wrong. This is because hooks can override the default listenerOptions saved in state, and when they do so by calling addMulticallListeners in slice.ts, the preferred blocksPerFetch is saved in callListeners in state. So activeListeningKeys and outdatedListeningKeys should be getting blocksPerFetch from callListeners in state, not the listenerOptions passed in thru Updater props.

File: src/updater.tsx (hunk 13 above)
Comment (lynnshaoyu) [reply src/updater.tsx:134]: Responded to in prev comment.

**Review by JFrankfurt — APPROVED** [review]
